### PR TITLE
contract on mainnet that only executes proposals

### PR DESCRIPTION
**Pull request by Hecate2**

### CODE REVIEW

**Comment by vang1ong7ang** [comment]
things to do:

- time lock
- emergency pause

File: NeoBurgerGovernanceToken.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        public static object ExecuteProposal(UInt160 scripthash, string method, ByteString[] args)
+        {
+            Runtime.CheckWitness((UInt160)Storage.Get(Storage.CurrentContext, new byte[] { PREFIX_TEE }));
+            return Contract.Call(scripthash, method, CallFlags.All, args);
+        }
+
+        public static void ChangeTEE(UInt160 newTEE)
+        {
+            ByteString new_tee_bytearray = (ByteString)new byte[] { PREFIX_TEE };
+            Runtime.CheckWitness((UInt160)Storage.Get(Storage.CurrentContext, new_tee_bytearray));
```
Comment (vang1ong7ang) [inline NeoBurgerGovernanceToken.cs]: add executionengine.assert here

File: NeoBurgerGovernanceToken.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+        public static object ExecuteProposal(UInt160 scripthash, string method, ByteString[] args)
+        {
+            Runtime.CheckWitness((UInt160)Storage.Get(Storage.CurrentContext, new byte[] { PREFIX_TEE }));
+            return Contract.Call(scripthash, method, CallFlags.All, args);
+        }
+
+        public static void ChangeTEE(UInt160 newTEE)
+        {
+            ByteString new_tee_bytearray = (ByteString)new byte[] { PREFIX_TEE };
```
Comment (vang1ong7ang) [inline NeoBurgerGovernanceToken.cs]: we use one-word var name to keep this coding style

File: NeoBurgerTestnetVote.cs [hunk 3]
```diff
@@ -0,0 +1,155 @@
+﻿using System;
```
Comment (vang1ong7ang) [inline NeoBurgerTestnetVote.cs]: move this file into another PR coz this file is less important

File: NeoBurgerGovernanceToken.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    [ManifestExtra("Description", "NeoBurger Governance Token")]
+    [SupportedStandards("NEP-17")]
+    [ContractPermission("*", "*")]
+    public class NeoBurgerGovernanceToken : Nep17Token
+    {
+        private const byte PREFIX_TEE = 0x50;
+        public override byte Decimals() => 8;
+        public override string Symbol() => "NOBUG";
+
+        public static UInt160 GetTEE() => (UInt160)Storage.Get(Storage.CurrentContext, new byte[] { PREFIX_TEE });
```
Comment (vang1ong7ang) [inline NeoBurgerGovernanceToken.cs]: change `GetTEE` to `TEE` for coding style

File: NeoBurgerGovernanceToken.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            if ((BigInteger)Storage.Get(Storage.CurrentContext, (ByteString)new byte[] { PREFIX_PAUSED } + (ByteString)proposal_id) > 0)
+                throw new System.Exception("proposal paused");
+            Storage.Delete(Storage.CurrentContext, proposalPassedKey);
+            return Contract.Call(scripthash, method, CallFlags.All, args);
+        }
+
+        public static void ChangeTEE(UInt160 newTEE)
+        {
+            ByteString newTEEBytearray = (ByteString)new byte[] { PREFIX_TEE };
+            ExecutionEngine.Assert(Runtime.CheckWitness((UInt160)Storage.Get(Storage.CurrentContext, newTEEBytearray)));
```
Comment (vang1ong7ang) [inline NeoBurgerGovernanceToken.cs]: use `ExecutionEngine.Assert(Runtime.CheckWitness(Runtime.ExecutingScriptHash));` instead to ensure the timelock

File: NeoBurgerGovernanceToken.cs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+            // TODO: MIN(OLD, NEW)
+            Storage.Put(Storage.CurrentContext, new byte[] { PREFIX_PAUSEUNTIL }, newPauseUntil);
+        }
+
+        public static object Execute(UInt160 scripthash, string method, object[] args)
+        {
+            ExecutionEngine.Assert(Proceed());
+            // TODO: CALCULATE DIGEST
+            BigInteger timestamp = (BigInteger)Storage.Get(Storage.CurrentContext, new byte[] { PREFIX_EXECUTION } + digest);
+            ExecutionEngine.Assert(timestamp + DEFAULT_WAITTIME > Runtime.Time);
```
Comment (vang1ong7ang) [inline NeoBurgerGovernanceToken.cs]: storage delete

File: NeoBurgerGovernanceToken.cs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+        public static void _deploy(object data, bool update)
+        {
+            Storage.Put(Storage.CurrentContext, new byte[] { PREFIX_TEE }, DEFAULT_TEE);
+        }
+
+
+        public static void SubmitApprovedExecution(UInt256 digest)
+        {
+            ExecutionEngine.Assert(Runtime.CheckWitness(TEE()));
```
Comment (vang1ong7ang) [inline NeoBurgerGovernanceToken.cs]: assert Proceed()

File: NeoBurgerGovernanceToken.cs [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+                newPauseUntil = currentPauseUntil;
+            else
+                throw new Exception("Cannot pause the contract longer");
+            Storage.Put(Storage.CurrentContext, new byte[] { PREFIX_PAUSEUNTIL }, newPauseUntil);
+        }
+
+        public static object Execute(BigInteger id, UInt160 scripthash, string method, object[] args)
+        {
+            ExecutionEngine.Assert(NotPaused());
+            ByteString digest = CryptoLib.Sha256(StdLib.Serialize(id) + StdLib.Serialize(scripthash) + StdLib.Serialize(method) + StdLib.Serialize(args));
```
Comment (vang1ong7ang) [inline NeoBurgerGovernanceToken.cs]: aggrate the 3 object into an array and then serialize

File: NeoBurgerGovernanceToken.cs [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+            StorageMap executionSubmittedTimeMap = new(Storage.CurrentContext, PREFIX_EXECUTION);
+            BigInteger timestamp = (BigInteger)executionSubmittedTimeMap.Get(digest);
+            BigInteger currentTime = Runtime.Time;
+            if (timestamp + DEFAULT_WAITTIME > currentTime)
```
Comment (vang1ong7ang) [inline NeoBurgerGovernanceToken.cs]: just use `Assert` without `try` and `throw`

File: NeoBurgerGovernanceToken.cs [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        }
+
+        public static void SubmitExecution(UInt256 digest)
+        {
+            ExecutionEngine.Assert(BalanceOf(Runtime.CallingScriptHash) > TotalSupply() / 2);
+            StorageMap executionSubmittedTimeMap = new(Storage.CurrentContext, PREFIX_EXECUTION);
+            executionSubmittedTimeMap.Put(digest, Runtime.Time);
+        }
+
+        public static void PauseContract()
```
Comment (vang1ong7ang) [inline NeoBurgerGovernanceToken.cs]: move this to onnep17 payment

**Comment by vang1ong7ang** [comment]
The Merkle hash root does not indicate the tree depth, enabling a second-preimage attack in which an attacker creates a document other than the original that has the same Merkle hash root. For the example above, an attacker can create a new document containing two data blocks, where the first is hash 0-0 + hash 0–1, and the second is hash 1-0 + hash 1-1.[13][14]

One simple fix is defined in Certificate Transparency: when computing leaf node hashes, a 0x00 byte is prepended to the hash data, while 0x01 is prepended when computing internal node hashes.[12] Limiting the hash tree size is a prerequisite of some formal security proofs, and helps in making some proofs tighter. Some implementations limit the tree depth using hash tree depth prefixes before hashes, so any extracted hash chain is defined to be valid only if the prefix decreases at each step and is still positive when the leaf is reached.

**Comment by Hecate2** [comment]
The argument `UInt256[] proof` in method `claim` should have hashes closer to the root with larger indices
